File: app/src/main/java/blue/koenig/kingsfinances/features/statistics/FinanceAssetsCalculatorService.java
package blue.koenig.kingsfinances.features.statistics;

import android.content.Context;

import com.koenig.commonModel.Byteable;
import com.koenig.commonModel.finance.BankAccount;

import org.joda.time.DateTime;

import java.nio.ByteBuffer;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import blue.koenig.kingsfamilylibrary.model.FamilyConfig;
import blue.koenig.kingsfinances.R;
import blue.koenig.kingsfinances.model.calculation.StatisticEntry;

/**
 * Created by Thomas on 07.01.2018.
 */

public class FinanceAssetsCalculatorService implements AssetsCalculatorService {
    private static final String ASSETS = "ASSETS";

    Context context;
    private DateTime startDate;

    public FinanceAssetsCalculatorService(Context context, DateTime startDate) {
        this.context = context;
        this.startDate = startDate;
    }

    @Override
    public Map<BankAccount, List<StatisticEntry>> loadAllBankAccountStatistics() {
        ByteBuffer buffer = FamilyConfig.getBytesFromConfig(context, ASSETS);
        if (buffer == null) return new HashMap<>();

        int size = buffer.getInt();
        Map<BankAccount, List<StatisticEntry>> listMap = new HashMap<>(size);
        for (int i = 0; i < size; i++) {
            BankAccount bankAccount = new BankAccount(buffer);
            int entries = buffer.getInt();
            List<StatisticEntry> statistics = new ArrayList<>(entries);
            for (int j = 0; j < entries; j++) {
                statistics.add(new StatisticEntry(buffer));
            }

            listMap.put(bankAccount, statistics);
        }

        return listMap;
    }

    @Override
    public DateTime getStartDate() {
        // TODO: preferences or somewhere
        return startDate;
    }

    @Override
    public DateTime getEndDate() {
        return DateTime.now();
    }

    @Override
    public String getOverallString() {
        return context.getResources().getString(R.string.overall);
    }

    @Override
    public void save(Map<BankAccount, List<StatisticEntry>> statisticEntryLists) {
        int size = 4;
        for (BankAccount bankAccount : statisticEntryLists.keySet()) {
            size += bankAccount.getByteLength();
            size += Byteable.getBigListLength(statisticEntryLists.get(bankAccount));
        }

        ByteBuffer buffer = ByteBuffer.allocate(size);
        buffer.putInt(statisticEntryLists.size());
        for (BankAccount bankAccount : statisticEntryLists.keySet()) {
            bankAccount.writeBytes(buffer);
            Byteable.writeBigList(statisticEntryLists.get(bankAccount), buffer);
        }

        FamilyConfig.saveBytes(context, buffer.array(), ASSETS);
    }

    @Override
    public String getFutureString() {
        return context.getString(R.string.future);
    }
}

File: app/src/main/java/blue/koenig/kingsfinances/features/statistics/FinanceAssetsCalculatorService.kt
package blue.koenig.kingsfinances.features.statistics

import blue.koenig.kingsfinances.model.FinanceConfig
import blue.koenig.kingsfinances.model.calculation.StatisticEntryDeprecated
import com.koenig.commonModel.Byteable
import com.koenig.commonModel.finance.BankAccount
import org.joda.time.DateTime
import java.nio.ByteBuffer
import java.util.*

/**
 * Created by Thomas on 07.01.2018.
 */

class FinanceAssetsCalculatorService(private val config: FinanceConfig) : AssetsCalculatorService {
    override val overallString: String
        get() = config.overallString

    override val futureString: String
        get() = config.futureString
    override lateinit var startDate: DateTime
    override val endDate: DateTime = DateTime.now()


    init {
        config.startDateObservable.subscribe { startDate = it }
    }

    override fun loadAllBankAccountStatistics(): Map<BankAccount, List<StatisticEntryDeprecated>> {
        val buffer = config.loadBuffer(ASSETS) ?: return HashMap()

        val size = buffer.int
        val listMap = HashMap<BankAccount, List<StatisticEntryDeprecated>>(size)
        for (i in 0 until size) {
            val bankAccount = BankAccount(buffer)
            val entries = buffer.int
            val statistics = ArrayList<StatisticEntryDeprecated>(entries)
            for (j in 0 until entries) {
                statistics.add(StatisticEntryDeprecated(buffer))
            }

            listMap.put(bankAccount, statistics)
        }

        return listMap
    }


    override fun save(statisticEntryLists: Map<BankAccount, List<StatisticEntryDeprecated>>) {
        var size = 4
        for (bankAccount in statisticEntryLists.keys) {
            size += bankAccount.byteLength
            size += Byteable.getBigListLength(statisticEntryLists[bankAccount]!!)
        }

        val buffer = ByteBuffer.allocate(size)
        buffer.putInt(statisticEntryLists.size)
        for (bankAccount in statisticEntryLists.keys) {
            bankAccount.writeBytes(buffer)
            Byteable.writeBigList(statisticEntryLists[bankAccount]!!, buffer)
        }

        config.saveBuffer(buffer, ASSETS)
    }


    companion object {
        private val ASSETS = "ASSETS"
    }
}
